astropy/utils/data.py

```diff
@@ -977,145 +1000,148 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
 
     if timeout is None:
         timeout = conf.remote_timeout
+    if sources is None:
+        sources = [remote_url]
+    if not sources:
+        raise ValueError(
+            "No sources listed! Please include primary URL if you want it "
+            "to be included as a valid source.")
+    if update_cache and not cache:
+        raise ValueError("update_cache only makes sense when using the cache!")
 
     missing_cache = False
 
-    if cache:
-        try:
-            dldir, urlmapfn = _get_download_cache_locs()
-        except OSError as e:
-            msg = 'Remote data cache could not be accessed due to '
-            estr = '' if len(e.args) < 1 else (': ' + str(e))
-            warn(CacheMissingWarning(msg + e.__class__.__name__ + estr))
-            cache = False
-            missing_cache = True  # indicates that the cache is missing to raise a warning later
-
     url_key = remote_url
 
-    # Check if URL is Astropy data server, which has alias, and cache it.
-    if (url_key.startswith(conf.dataurl) and
-            conf.dataurl not in _dataurls_to_alias):
-        try:
-            with urllib.request.urlopen(conf.dataurl, timeout=timeout) as remote:
-                _dataurls_to_alias[conf.dataurl] = [conf.dataurl, remote.geturl()]
-        except urllib.error.URLError:  # Host unreachable
-            _dataurls_to_alias[conf.dataurl] = [conf.dataurl]
-    try:
-        if cache:
-            # We don't need to acquire the lock here, since we are only reading
-            with shelve.open(urlmapfn) as url2hash:
-                if url_key in url2hash:
+    if cache:
+        with _cache() as (dldir, url2hash):
+            if dldir is None:
+                cache = False
+                update_cache = False
+                missing_cache = True  # emit a warning later
+            elif update_cache:
+                # just check the cache works but we want a new value
+                pass
+            else:
+                try:
                     return url2hash[url_key]
-                # If there is a cached copy from mirror, use it.
-                else:
-                    for cur_url in _dataurls_to_alias.get(conf.dataurl, []):
-                        if url_key.startswith(cur_url):
-                            url_mirror = url_key.replace(cur_url,
-                                                         conf.dataurl_mirror)
-                            if url_mirror in url2hash:
-                                return url2hash[url_mirror]
-
-        with urllib.request.urlopen(remote_url, timeout=timeout) as remote:
-            # keep a hash to rename the local file to the hashed name
-            hash = hashlib.md5()
-
-            info = remote.info()
-            if 'Content-Length' in info:
+                except KeyError:
+                    pass
+
+    errors = {}
+    for source_url in sources:
+        try:
+            with urllib.request.urlopen(source_url, timeout=timeout) as remote:
```

Comment:
Should we be using [urlretrieve](https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve) instead, as suggested by @Cadair and @bsipocz before in a different issue? It already raises `ContentTooShortError` when download dies prematurely, thus further simplifying the implementation here.

The downside is that we cannot easily pass custom header with `urlretrieve` but do we need to do it? Ref: https://docs.python.org/3/howto/urllib2.html#headers

Reply:
That would probably work fine, I was not totally sure how to combine it with our progress bar. Shouldn't be a problem, and if the hash isn't easy to compute on the fly compute_hash is cheap and easy. There isn't a lot of practical difference apart from (I hope) cleaner python2 compatibility, less code on our part, and a shortcut that avoids copying local files.

Reply:
The key problem is that the progress thing needs to know the size of the file when initialized, and urlretrieve only reports this in its progress updates. I'm trying a hack involving contextlib.ExitStack, but I'm not sure what if any information you get if (say) DNS resolving takes a long time, or the server connects but never provides any content.

Reply:
Ack. Okay, the progress bar thing is solvable. Making sure the temporary gets deleted is annoying but doable. Unfortunately, urlretrieve does not support setting a timeout. Since we explicitly provide a timeout in the function signature, and there are timeout settings in the config file, I think this overrides some code simplification - especially if there's not actually any simplification. Curses, I had something written and mostly working so I went on and made other changes.

Reply:
> cleaner python2 compatibility

We don't need that compatibility here at all.

Reply:
Er. sorry, of course you're right. Thoughts crossing from another project. But Windows often raises different exceptions and it would be nice to rely on urlretrieve to handle them correctly. Unfortunately I concluded that the absence of a way to use timeouts ruled out urlretrieve. I *think* I've handled all the various Windows cases but it's a pain (and means a lot of tiny patches added to the PR).